Client/TabManagement/TabManagerImplementation.swift

```diff
@@ -97,8 +98,11 @@ class TabManagerImplementation: LegacyTabManager, Notifiable {
 
     private func buildTabRestore(window: WindowData?) async {
         defer {
-            isRestoringTabs = false
-            tabRestoreHasFinished = true
+            ensureMainThread { [weak self] in
```

Comment:
You shouldn't use GCD within an async context like this, it manually manages threads under the hood which you aren't supposed to do with swift concurrency. To achieve the same thing you can instead do 
`await MainActor.run { }` or `Task { @MainActor in }`

Reply:
This function is already using `ensureMainThread` for the delegate callbacks (a bit further down) but I don't have much context on why that was added. But one of the (other) problems with it is that it creates a race between those callbacks and the `defer` closure, depending on what thread this ends up being executed on. In some cases those callbacks may happen synchronously and in other cases they may be `dispatch_async`'d and the `defer` closure in that case will happen first.

In my initial reading of this the intent seems to be that the defer block is expected to happen last but it's not entirely clear from the code itself. This area could probably use a bit more clarification at some point; I'll remove this additional `ensureMainThread` in the meantime and can revisit this.

Reply:
Oops, that's my bad, I recently moved that code to there from somewhere else. I'll fix up the other ensureMainThread in another task.